Installation issue: boost on Ubuntu 20.04

<!-- Thanks for taking the time to report this build failure. To proceed with the report please:

1. Title the issue "Installation issue: <name-of-the-package>".
2. Provide the information required below.

We encourage you to try, as much as possible, to reduce your problem to the minimal example that still reproduces the issue. That would help us a lot in fixing it quickly and effectively! -->

### Steps to reproduce the issue
This is related to issue #22252. On Ubuntu 20.04, Spack fails to build 'boost', indicating errors during the install phase. Below is the sequence of steps to reproduce the error, which happens either when trying to build with the Ubuntu 20.04 default gcc 9 or if I first install gcc 8.3.0 within Spack 
<!-- Fill in the exact spec you are trying to build and the relevant part of the error message -->
```console
$ spack install gcc@8.3.0
$ spack load gcc@8.3.0
$ spack compiler find
$ spack install boost@1.63.0+python+numpy%gcc@8.3.0 ^py-numpy@1.19.5 ^python@3.6.13
==> boost: Executing phase: 'install'
==> Error: ProcessError: Command exited with status 1:
    './b2' 'install' 'threading=multi' '-j' '8' '--user-config=/tmp/apvieira/spack-stage/spack-stage-boost-1.63.0-mfubaxjxoj7z5bmgmmyy6qeeb3rgzj2p/spack-src/user-config.jam' 'variant=release' '-s' 'BZIP2_INCLUDE=/home/apvieira/spack/opt/spack/linux-ubuntu20.04-ivybridge/gcc-9.3.0/bzip2-1.0.8-m3hgpkjzvztgbqidh6rsmpqvmgwvs2sv/include' '-s' 'BZIP2_LIBPATH=/home/apvieira/spack/opt/spack/linux-ubuntu20.04-ivybridge/gcc-9.3.0/bzip2-1.0.8-m3hgpkjzvztgbqidh6rsmpqvmgwvs2sv/lib' '-s' 'ZLIB_INCLUDE=/home/apvieira/spack/opt/spack/linux-ubuntu20.04-ivybridge/gcc-9.3.0/zlib-1.2.11-f2quknrcbcclvrubs7rorxvtp5uxbq4d/include' '-s' 'ZLIB_LIBPATH=/home/apvieira/spack/opt/spack/linux-ubuntu20.04-ivybridge/gcc-9.3.0/zlib-1.2.11-f2quknrcbcclvrubs7rorxvtp5uxbq4d/lib' '-s' 'NO_LZMA=1' '-s' 'NO_ZSTD=1' 'link=static,shared' '--layout=system' 'toolset=gcc' 'cxxflags="-std=c++98"'

11962 errors found in build log:
     13731                     from /usr/include/unicode/unistr.h:39,
     13732                     from /usr/include/unicode/strenum.h:20,
     13733                     from /usr/include/unicode/locid.h:40,
     13734                     from /usr/include/unicode/coll.h:64,
     13735                     from ./boost/regex/icu.hpp:24,
     13736                     from libs/regex/build/../src/icu.cpp:23:
  >> 13737    /usr/include/c++/9/bits/c++0x_warning.h:32:2: error: #error This file requires compiler and library support for the ISO C++ 2011 standard. This support must be enabled with the -std=c++11 or -std=gnu++11 compiler options.
     13738       32 | #error This file requires compiler and library support \

...
```

### Information on your system

<!-- Please include the output of `spack debug report` -->
* **Spack:** 0.16.1-1698-d22bad13b1
* **Python:** 3.8.5
* **Platform:** linux-ubuntu20.04-ivybridge
* **Concretizer:** original
[spack-build-out.txt](https://github.com/spack/spack/files/6138099/spack-build-out.txt)
[spack-build-env.txt](https://github.com/spack/spack/files/6138102/spack-build-env.txt)

<!-- If you have any relevant configuration detail (custom `packages.yaml` or `modules.yaml`, etc.) you can add that here as well. -->

### Additional information

<!-- Please upload the following files. They should be present in the stage directory of the failing build. Also upload any config.log or similar file if one exists. -->
* [spack-build-out.txt]()
* [spack-build-env.txt]()

<!-- Some packages have maintainers who have volunteered to debug build failures. Run `spack maintainers <name-of-the-package>` and @mention them here if they exist. -->
@hainest
### General information

<!-- These boxes can be checked by replacing [ ] with [x] or by clicking them after submitting the issue. -->
- [x] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [x] I have run `spack maintainers <name-of-the-package>` and @mentioned any maintainers
- [x] I have uploaded the build log and environment files
- [x] I have searched the issues of this repo and believe this is not a duplicate

--- a/var/spack/repos/builtin/packages/boost/package.py
+++ b/var/spack/repos/builtin/packages/boost/package.py
@@ -159,7 +159,13 @@ class Boost(Package):
             description='Default symbol visibility in compiled libraries '
             '(1.69.0 or later)')
 
+    # Unicode support
     depends_on('icu4c', when='+icu')
+    depends_on('icu4c cxxstd=11', when='+icu cxxstd=11')
+    depends_on('icu4c cxxstd=14', when='+icu cxxstd=14')
+    depends_on('icu4c cxxstd=17', when='+icu cxxstd=17')
+    conflicts('cxxstd=98', when='+icu')  # Requires c++11 at least
+
     depends_on('python', when='+python')
     depends_on('mpi', when='+mpi')
     depends_on('bzip2', when='+iostreams')
@@ -331,6 +337,11 @@ class Boost(Package):
         if '+python' in spec:
             options.append('--with-python=%s' % spec['python'].command.path)
 
+        if '~icu' in spec:
+            options.append('--without-icu')
+        else:
+            options.append('--with-icu')
+
         with open('user-config.jam', 'w') as f:
             # Boost may end up using gcc even though clang+gfortran is set in
             # compilers.yaml. Make sure this does not happen:
@@ -366,8 +377,10 @@ class Boost(Package):
         else:
             options.append('variant=release')
 
-        if '+icu_support' in spec:
-            options.extend(['-s', 'ICU_PATH=%s' % spec['icu'].prefix])
+        if '+icu' in spec:
+            options.extend(['-s', 'ICU_PATH=%s' % spec['icu4c'].prefix])
+        else:
+            options.append('--disable-icu')
 
         if '+iostreams' in spec:
             options.extend([

--- a/var/spack/repos/builtin/packages/boost/package.py
+++ b/var/spack/repos/builtin/packages/boost/package.py
@@ -337,10 +337,10 @@ class Boost(Package):
         if '+python' in spec:
             options.append('--with-python=%s' % spec['python'].command.path)
 
-        if '~icu' in spec:
-            options.append('--without-icu')
-        else:
+        if '+icu' in spec:
             options.append('--with-icu')
+        else:
+            options.append('--without-icu')
 
         with open('user-config.jam', 'w') as f:
             # Boost may end up using gcc even though clang+gfortran is set in

